<a href="https://colab.research.google.com/github/PosgradoMNA/Proyecto-final-Equipo-135/blob/main/Ciencia_de_Datos_Semana_3_Actividad_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Parte 1: Fundamentos de bases de datos**





1.  ** Fundamentos de bases de datos y bases para ciencia de datos.**

*Para mi, mi primer contacto fue con los documentos de Excel, como bien son citados los documentos tienen cierta base "estructurada" que los permite ser manejados de una manera mucho mas eficiente, pero en muchos casos no le quita lo complicado al buscar, filtrar, remplazar, realizar operaciones, etcetera...*

  Una base de datos es un repositorio donde se almacenan diferentes tipos de datos provenientes de diferentes fuentes con el propósito de tener un rápido acceso a estos de manera seleccionada para cualquier tipo de análisis que sea requerido. Una característica de las bases de datos es que no cuentan con herramientas propias para realizar análisis en su mismo entorno, sino que los datos se deben extraer para después poderse procesar externamente, lo cual permite que su enfoque sea únicamente el de almacenar y facilitar la extracción de la información de la manera más eficiente posible y se asegura como resultado la calidad y consistencia de la misma.

  En ciencia de datos, es una tarea común la de localizar y recopilar datos de diferentes fuentes y bases de datos y reorganizarla en un conjunto de datos que sean relevantes para el caso de estudio que se está tratando y que sirva para contestar preguntas que nos lleven a los hallazgos con los que podemos fundamentar la toma de decisiones en una problemática específica del negocio. En otras ocasiones los datos puede ser que vengan de una sola base de datos, pero en un caso real no es lo más común, por la complejidad de variables que intervienen y que se desean medir.

  Una de las diferencias entre un analista de datos y un científico de datos es la habilidad de poder predecir datos en el futuro por medio de modelos que mejoran con el tiempo, mientras que un analista sólo observa datos históricos y la toma de decisiones basada en estos datos no le permite a la organización anticiparse a cambios en los datos reales y tener una ventaja competitiva con respecto a sus competidores. 


2.   **Fundamentos de almacenes de datos (Data Warehouse) para ciencia de datos.**

*Conforme fui intentando desarrollar herramientas que me permitieran simplificar mi trabajo, descubri poco a poco la parte de la programación, para poder acomodar, eliminar errores (como los campos vacios al intentar capturar información) y muchas mejoras mas, pero despues descubrí MS Access, me ayudo a simplificar incluso mucho mas con las API que podrían desarrollarse para hacer mucho mas amigable e incluso eficiente la captura de información, poco despues comence a saber sobre estrucuras mucho mas "profesionales" como SQL, MySQL y Oracle aunque estas solo las llegue a utilizar para poder comprender la forma en la que trabajaban ya que fui entrenado para poder utilizar una herramienta que se llama Crystal Reports con la que comenzamos a hacer reportes para todos los departamentos de la compañía para la que trabajaba*

  Un almacén de datos contiene información de muchas fuentes diferentes y realiza una serie de transformaciones a los datos de entrada para permitirnos hacer búsquedas que abarcan varias de esas fuentes que lo alimentan, por lo que es común que sean utilizados por las grandes compañías transnacionales que por su naturaleza generan datos constantemente de una manera masiva.

  Por otro lado, la diversidad en la información que almacenan implica una complejidad significativa en cuanto a su manejo, ya que al no haber un filtro en la alimentación, es necesario implementar mecanismos que nos ayuden a determinar si existe o no relación entre los diferentes tipos de información, así como asegurar que los datos faltantes o erróneos se identifiquen y reemplacen, realizar transformaciones para hacer que los datos sean comparables entre sí, lo que se conoce como normalización, etc. 


# **Parte 2: Selección y limpieza de los Datos en Python**

Para esta parte, utilizaremos un set de datos ubicado en la siguiente dirección:
https://raw.githubusercontent.com/PosgradoMNA/Actividades_Aprendizaje-/main/default%20of%20credit%20card%20clients.csv

*esta información será accedida a travez de Pandas (PD)
Para poder mandar a llamar la base de datos primero tenemos que establecer un path de tal forma que Python/Pandas sepa donde se encuentra el documento con el que estara trabajando.*

*y sera con las siguientes lineas a continuación:*


In [15]:
import pandas as pd

data = pd.read_csv('https://raw.githubusercontent.com/PosgradoMNA/Actividades_Aprendizaje-/main/default%20of%20credit%20card%20clients.csv')
print(data)

          ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20 

In [16]:
data.shape #nos explica como es la forma de nuestro Dataset

(30000, 25)

In [18]:
data.info() #Desplegamos un resumen general del data frame "data"

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   ID      30000 non-null  int64  
 1   X1      30000 non-null  int64  
 2   X2      29999 non-null  float64
 3   X3      29998 non-null  float64
 4   X4      29998 non-null  float64
 5   X5      29995 non-null  float64
 6   X6      29997 non-null  float64
 7   X7      29995 non-null  float64
 8   X8      29993 non-null  float64
 9   X9      29991 non-null  float64
 10  X10     29984 non-null  float64
 11  X11     29986 non-null  float64
 12  X12     29989 non-null  float64
 13  X13     29989 non-null  float64
 14  X14     29987 non-null  float64
 15  X15     29985 non-null  float64
 16  X16     29983 non-null  float64
 17  X17     29990 non-null  float64
 18  X18     29992 non-null  float64
 19  X19     29991 non-null  float64
 20  X20     29992 non-null  float64
 21  X21     29989 non-null  float64
 22

In [13]:
data.describe()   #Para tener un resumen estadístico de los datos

ID              X1            X2            X3            X4  \
count  29958.000000    29958.000000  29958.000000  29958.000000  29958.000000   
mean   15005.550504   167555.900928      1.604012      1.853094      1.551739   
std     8654.547473   129737.299088      0.489070      0.790471      0.521952   
min        1.000000    10000.000000      1.000000      0.000000      0.000000   
25%     7516.250000    50000.000000      1.000000      1.000000      1.000000   
50%    15005.500000   140000.000000      2.000000      2.000000      2.000000   
75%    22497.750000   240000.000000      2.000000      2.000000      2.000000   
max    30000.000000  1000000.000000      2.000000      6.000000      3.000000   

                 X5            X6            X7            X8            X9  \
count  29958.000000  29958.000000  29958.000000  29958.000000  29958.000000   
mean      35.483443     -0.017124     -0.134021     -0.166767     -0.221110   
std        9.214319      1.123989      1.197171      1.196026      1.168419   
min       21.000000     -2.000000     -2.000000     -2.000000     -2.000000   
25%       28.000000     -1.000000     -1.000000     -1.000000     -1.000000   
50%       34.000000      0.000000      0.000000      0.000000      0.000000   
75%       41.000000      0.000000      0.000000      0.000000      0.000000   
max       79.000000      8.000000      8.000000      8.000000      8.000000   

       ...            X15            X16            X17            X18  \
count  ...   29958.000000   29958.000000   29958.000000   29958.000000   
mean   ...   43279.335370   40328.984578   38889.925763    5664.614460   
std    ...   64364.684347   60826.219326   59582.883301   16568.823518   
min    ... -170000.000000  -81334.000000 -339603.000000       0.000000   
25%    ...    2327.500000    1762.250000    1256.000000    1000.000000   
50%    ...   19037.500000   18104.500000   17067.500000    2100.000000   
75%    ...   54551.250000   50220.750000   49234.750000    5007.000000   
max    ...  891586.000000  927171.000000  961664.000000  873552.000000   

                X19            X20            X21            X22  \
count  2.995800e+04   29958.000000   29958.000000   29958.000000   
mean   5.925715e+03    5228.429969    4829.873556    4801.481574   
std    2.305598e+04   17617.338167   15676.205514   15285.552652   
min    0.000000e+00       0.000000       0.000000       0.000000   
25%    8.352500e+02     390.000000     296.250000     253.250000   
50%    2.009000e+03    1800.000000    1500.000000    1500.000000   
75%    5.000000e+03    4511.500000    4014.750000    4040.000000   
max    1.684259e+06  896040.000000  621000.000000  426529.000000   

                 X23             Y  
count   29958.000000  29958.000000  
mean     5220.708025      0.221143  
std     17788.983767      0.415023  
min         0.000000      0.000000  
25%       118.000000      0.000000  
50%      1500.000000      0.000000  
75%      4000.000000      0.000000  
max    528666.000000      1.000000  

[8 rows x 25 columns]

In [4]:
data.isnull().values.any() #nos dice si existen valores vacios

True

In [5]:
data.isna().values.any() #nos dice si existen valores NA

True

In [40]:
data.isnull().any() #nos dice si existen valores vacios en cada columna

ID     False
X1     False
X2      True
X3      True
X4      True
X5      True
X6      True
X7      True
X8      True
X9      True
X10     True
X11     True
X12     True
X13     True
X14     True
X15     True
X16     True
X17     True
X18     True
X19     True
X20     True
X21     True
X22     True
X23     True
Y       True
dtype: bool

In [73]:
#iteramos para rellenar campos vacíos en columnas 6 a 11 con la moda o el valor más frecuente por columna
#estos valores pueden van de -1 a 9 y reflejan morosidad de los clientes

for i in range(6,12):   
  data.iloc[:,i].fillna(data.iloc[:,i].mode, inplace=True)

In [74]:
data.isnull().any()   #Revisamos que ya no haya campos vacíos para variables X6 a X11

ID     False
X1     False
X2      True
X3      True
X4      True
X5      True
X6     False
X7     False
X8     False
X9     False
X10    False
X11    False
X12     True
X13     True
X14     True
X15     True
X16     True
X17     True
X18     True
X19     True
X20     True
X21     True
X22     True
X23     True
Y       True
dtype: bool

In [75]:
data.dropna(inplace = True) # Asigna en los lugares vacios un "True" para remover los campos

In [76]:
data # despliega el Dataset

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20     X21      X22     X23  \
0          0.0      0.0      0.0    689.0      0.0     0.0      0.0     0.0   
1       3455.0   3261.0      0.0   1000.0   1000.0  1000.0      0.0  2000.0   
2      14948.0  15549.0   1518.0   1500.0   1000.0  1000.0   1000.0  5000.0   
3      28959.0  29547.0   2000.0   2019.0   1200.0  1100.0   1069.0  1000.0   
4      19146.0  19131.0   2000.0  36681.0  10000.0  9000.0    689.0   679.0   
...        ...      ...      ...      ...      ...     ...      ...     ...   
29995  31237.0  15980.0   8500.0  20000.0   5003.0  3047.0   5000.0  1000.0   
29996   5190.0      0.0   1837.0   3526.0   8998.0   129.0      0.0     0.0   
29997  20582.0  19357.0      0.0      0.0  22000.0  4200.0   2000.0  3100.0   
29998  11855.0  48944.0  85900.0   3409.0   1178.0  1926.0  52964.0  1804.0   
29999  32428.0  15313.0   2078.0   1800.0   1430.0  1000.0   1000.0  1000.0   

         Y  
0      1.0  
1      1.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
29995  0.0  
29996  0.0  
29997  1.0  
29998  1.0  
29999  1.0  

[29971 rows x 25 columns]

Como podemos ver, ha cambiado el valor de los 30,000 a 29,971

In [77]:
data.isna().values.any() #muestra si hay valores con NA

False

In [78]:
data.isnull().values.any() #Muestra si hay valores vacios o "Empty"

False

In [79]:
data.sort_index(ascending=True) #realiza un sorteo d elos datos de forma acendente

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15  \
0          1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0   
1          2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   
2          3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0   
3          4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0   
4          5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0   
...      ...     ...  ...  ...  ...   ...  ...  ...  ...  ...  ...      ...   
29995  29996  220000  1.0  3.0  1.0  39.0  0.0  0.0  0.0  0.0  ...  88004.0   
29996  29997  150000  1.0  3.0  2.0  43.0 -1.0 -1.0 -1.0 -1.0  ...   8979.0   
29997  29998   30000  1.0  2.0  2.0  37.0  4.0  3.0  2.0 -1.0  ...  20878.0   
29998  29999   80000  1.0  3.0  1.0  41.0  1.0 -1.0  0.0  0.0  ...  52774.0   
29999  30000   50000  1.0  2.0  1.0  46.0  0.0  0.0  0.0  0.0  ...  36535.0   

           X16      X17      X18      X19      X20     X21      X22     X23  \
0          0.0      0.0      0.0    689.0      0.0     0.0      0.0     0.0   
1       3455.0   3261.0      0.0   1000.0   1000.0  1000.0      0.0  2000.0   
2      14948.0  15549.0   1518.0   1500.0   1000.0  1000.0   1000.0  5000.0   
3      28959.0  29547.0   2000.0   2019.0   1200.0  1100.0   1069.0  1000.0   
4      19146.0  19131.0   2000.0  36681.0  10000.0  9000.0    689.0   679.0   
...        ...      ...      ...      ...      ...     ...      ...     ...   
29995  31237.0  15980.0   8500.0  20000.0   5003.0  3047.0   5000.0  1000.0   
29996   5190.0      0.0   1837.0   3526.0   8998.0   129.0      0.0     0.0   
29997  20582.0  19357.0      0.0      0.0  22000.0  4200.0   2000.0  3100.0   
29998  11855.0  48944.0  85900.0   3409.0   1178.0  1926.0  52964.0  1804.0   
29999  32428.0  15313.0   2078.0   1800.0   1430.0  1000.0   1000.0  1000.0   

         Y  
0      1.0  
1      1.0  
2      0.0  
3      0.0  
4      0.0  
...    ...  
29995  0.0  
29996  0.0  
29997  1.0  
29998  1.0  
29999  1.0  

[29971 rows x 25 columns]

In [80]:
data.head() #muestra unicamente los encabezados (las 5 primeras filas)

ID      X1   X2   X3   X4    X5   X6   X7   X8   X9  ...      X15      X16  \
0   1   20000  2.0  2.0  1.0  24.0  2.0  2.0 -1.0 -1.0  ...      0.0      0.0   
1   2  120000  2.0  2.0  2.0  26.0 -1.0  2.0  0.0  0.0  ...   3272.0   3455.0   
2   3   90000  2.0  2.0  2.0  34.0  0.0  0.0  0.0  0.0  ...  14331.0  14948.0   
3   4   50000  2.0  2.0  1.0  37.0  0.0  0.0  0.0  0.0  ...  28314.0  28959.0   
4   5   50000  1.0  2.0  1.0  57.0 -1.0  0.0 -1.0  0.0  ...  20940.0  19146.0   

       X17     X18      X19      X20     X21     X22     X23    Y  
0      0.0     0.0    689.0      0.0     0.0     0.0     0.0  1.0  
1   3261.0     0.0   1000.0   1000.0  1000.0     0.0  2000.0  1.0  
2  15549.0  1518.0   1500.0   1000.0  1000.0  1000.0  5000.0  0.0  
3  29547.0  2000.0   2019.0   1200.0  1100.0  1069.0  1000.0  0.0  
4  19131.0  2000.0  36681.0  10000.0  9000.0   689.0   679.0  0.0  

[5 rows x 25 columns]

# **Parte 3: Preparación de los datos**




---


1.   ¿Qué datos considero mas importantes? 
**En realidad, podríamos considerar todos los datos pueden ser utiles en algun punto de nuestro analisis** *(a menos de que sean campos de texto o "strings" que no sean concretos, pero aun así en algun momento podríamos estar buscando palabras claves en estos campos por lo que, en general no podríamos descartar tan facilmente los campos)*
**Para este caso, los campos que consideramos que podrían ser importantes serían los siguientes: **

| Variables consideradas                    | Razón por la que se consideraron    | 
| ------------------------------- | ------------------------ |
| X1  = Amount of the given credit  | Tener los montos totales de credito otorgados  |
| X4  = Marital status  | Entender si cuenta con dependientes  |
| X5  = Age (year).   | Capacidad de pago / Trabajo  |
| X6  = the repayment status in Sep, 2005  | Desarrollo de estadisticas y status de pagos  |
| X7  = the repayment status in Aug, 2005  | 
| X8  = the repayment status in Jul, 2005  | 
| X9  = the repayment status in Jun, 2005  | 
| X10 = the repayment status in May, 2005  | 
| X11 = the repayment status in Apr, 2005  | 
| X12 = amount of bill statement in Sep, 2005  | Para el desarrollo de histogramas de pagos  |
| X13 = amount of bill statement in Aug, 2005  | 
| X14 = amount of bill statement in Jul, 2005  | 
| X15 = amount of bill statement in Jun, 2005  | 
| X16 = amount of bill statement in May, 2005  | 
| X17 = amount of bill statement in Apr, 2005  | 
| X18 = amount paid in September, 2005  | Para el desarrollo de histogramas y modelos de pago por montos  |
| X19 = amount paid in September, 2005  | 
| X20 = amount paid in September, 2005  | 
| X21 = amount paid in September, 2005  | 
| X22 = amount paid in September, 2005  | 
| X23 = amount paid in September, 2005  | 
 
 **Lista de variables no consideradas como "importantes**


| Variables excluidas       | Razón de exclusión    | 
| --- | --- |
| X2 = Gender | El genero solo podría aportar informacion estadística, no para la resolución de pagos  |
| X3 = Education  | Solo funcionaría con fines estadisticos ya que no esta directamente relacionada a la capacidad de pagos de creditos  |



2.   ¿Por qué?
**Consideramos que lo más importante es primero definir que información será la que se utilizará para poder partir con dicha información y el resto poder discriminarla** *(incluso por la optimización de los procesos y velocidad de manejo de información)*



3.   ¿Se eliminaron o reemplazaron datos nulos?
**Cuando estos pudieron ser remplazados, fueron remplazados, cuando no, fueron eliminados de tal forma que no contaminaran la información colectada**

4.   ¿Qué se hizo y por qué?

5.   ¿Es necesario ordenar los datos para el análisis? Sí / No / 
**No es necesario, pero se realizo por "estética" y para asegurar que toda la información estuviera organizada de forma Ascendente**
6.   ¿Por qué?

7.   ¿Existen problemas de formato que deban solucionar antes del proceso de 
modelado? Sí / No / 
**SI**
8.   Por qué.
**Por citar algunos ejemplos, podemos tener campos de fechas que no concuerden con lo que estaremos buscando, por ejemplo AAAA-MM-DD, si fue capturado como AAAA-DD-MM esto, en nuestro analizis nos traera inconvenientes y errores en el despliegue de información correcta**
9.   ¿Qué ajustes se realizaron en el proceso de limpieza de datos (agregar, integrar, eliminar, modificar registros (filas), cambiar atributos (columnas)?
**Para este Dataset se realizaron revisiónes de los campos vaciós, se rellenaron con las medias de los datos para "estandarizarlo", y no perder información que podría ser confiable para la entrega de nuestros resultados.**

## *A continuación mostraríamos los resultados tras los procesamientos de la información con los analisis previamente expuestos:*
